File: autonomica-api/run_integration_tests.py
#!/usr/bin/env python3
"""
Integration Test Runner for Autonomica API

This script runs integration tests with proper environment setup,
database initialization, and cleanup.
"""

import os
import sys
import subprocess
import tempfile
import shutil
import time
import signal
from pathlib import Path

def setup_test_environment():
    """Setup test environment variables and temporary directories."""
    print("🔧 Setting up test environment...")
    
    # Create temporary test directory
    test_dir = tempfile.mkdtemp(prefix="autonomica_test_")
    os.environ["TEST_DIR"] = test_dir
    
    # Set test environment variables
    test_env = {
        "TESTING": "true",
        "ENVIRONMENT": "test",
        "LOG_LEVEL": "DEBUG",
        "REDIS_URL": "redis://localhost:6379/1",
        "DATABASE_URL": f"sqlite:///{test_dir}/test.db",
        "OPENAI_API_KEY": "test-integration-key",
        "CLERK_SECRET_KEY": "test-integration-key",
        "AI_PROVIDER": "mock",
        "PYTHONPATH": str(Path(__file__).parent)
    }
    
    for key, value in test_env.items():
        os.environ[key] = value
        print(f"  {key}={value}")
    
    return test_dir

def check_dependencies():
    """Check if required dependencies are available."""
    print("🔍 Checking dependencies...")
    
    try:
        import pytest
        print(f"  ✅ pytest {pytest.__version__}")
    except ImportError:
        print("  ❌ pytest not found")
        return False
    
    try:
        import redis
        print(f"  ✅ redis {redis.__version__}")
    except ImportError:
        print("  ❌ redis not found")
        return False
    
    try:
        import fastapi
        print(f"  ✅ fastapi {fastapi.__version__}")
    except ImportError:
        print("  ❌ fastapi not found")
        return False
    
    return True

def start_redis_server():
    """Start Redis server for testing if not already running."""
    print("🚀 Starting Redis server...")
    
    try:
        import redis
        r = redis.Redis(host='localhost', port=6379, db=1)
        r.ping()
        print("  ✅ Redis server already running")
        return True
    except:
        print("  ⚠️  Redis server not running")
        print("  💡 Please start Redis server manually or use Docker:")
        print("     docker run -d -p 6379:6379 redis:7-alpine")
        return False

def run_integration_tests():
    """Run the integration test suite."""
    print("🧪 Running integration tests...")
    
    # Test configuration
    test_args = [
        "pytest",
        "tests/integration/",
        "-v",
        "--tb=short",
        "--cov=app",
        "--cov-report=term-missing",
        "--cov-report=html:htmlcov_integration",
        "--cov-report=xml:coverage_integration.xml",
        "--durations=10",
        "-m", "integration"
    ]
    
    print(f"  Running: {' '.join(test_args)}")
    
    try:
        result = subprocess.run(test_args, capture_output=False, text=True)
        return result.returncode == 0
    except Exception as e:
        print(f"  ❌ Error running tests: {e}")
        return False

def run_worker_integration_tests():
    """Run worker pod integration tests."""
    print("🏭 Running worker integration tests...")
    
    worker_test_dir = Path(__file__).parent.parent / "autonomica-worker" / "tests" / "integration"
    
    if not worker_test_dir.exists():
        print("  ⚠️  Worker test directory not found")
        return True
    
    test_args = [
        "pytest",
        str(worker_test_dir),
        "-v",
        "--tb=short",
        "--cov=.",
        "--cov-report=term-missing",
        "--cov-report=html:htmlcov_worker_integration",
        "--cov-report=xml:coverage_worker_integration.xml",
        "--durations=10",
        "-m", "integration"
    ]
    
    print(f"  Running: {' '.join(test_args)}")
    
    try:
        result = subprocess.run(test_args, capture_output=False, text=True)
        return result.returncode == 0
    except Exception as e:
        print(f"  ❌ Error running worker tests: {e}")
        return False

def cleanup_test_environment(test_dir):
    """Clean up test environment and temporary files."""
    print("🧹 Cleaning up test environment...")
    
    try:
        if os.path.exists(test_dir):
            shutil.rmtree(test_dir)
            print(f"  ✅ Removed test directory: {test_dir}")
    except Exception as e:
        print(f"  ⚠️  Error cleaning up test directory: {e}")
    
    # Clean up environment variables
    test_env_vars = [
        "TESTING", "ENVIRONMENT", "LOG_LEVEL", "REDIS_URL",
        "DATABASE_URL", "OPENAI_API_KEY", "CLERK_SECRET_KEY",
        "AI_PROVIDER", "TEST_DIR"
    ]
    
    for var in test_env_vars:
        if var in os.environ:
            del os.environ[var]

def generate_test_report():
    """Generate a summary test report."""
    print("\n📊 Integration Test Report")
    print("=" * 50)
    
    # Check for coverage reports
    coverage_files = [
        "htmlcov_integration",
        "coverage_integration.xml",
        "htmlcov_worker_integration",
        "coverage_worker_integration.xml"
    ]
    
    for file_path in coverage_files:
        if os.path.exists(file_path):
            print(f"  ✅ {file_path}")
        else:
            print(f"  ❌ {file_path} (not found)")
    
    print("\n📁 Test Results Location:")
    print("  - HTML Coverage: htmlcov_integration/")
    print("  - XML Coverage: coverage_integration.xml")
    print("  - Worker Coverage: htmlcov_worker_integration/")

def main():
    """Main integration test runner."""
    print("🚀 Autonomica Integration Test Runner")
    print("=" * 50)
    
    test_dir = None
    
    try:
        # Check dependencies
        if not check_dependencies():
            print("❌ Missing required dependencies")
            sys.exit(1)
        
        # Setup test environment
        test_dir = setup_test_environment()
        
        # Check Redis server
        if not start_redis_server():
            print("⚠️  Continuing without Redis server...")
        
        # Run API integration tests
        api_success = run_integration_tests()
        
        # Run worker integration tests
        worker_success = run_worker_integration_tests()
        
        # Generate report
        generate_test_report()
        
        # Determine overall success
        if api_success and worker_success:
            print("\n🎉 All integration tests completed successfully!")
            sys.exit(0)
        else:
            print("\n❌ Some integration tests failed!")
            sys.exit(1)
            
    except KeyboardInterrupt:
        print("\n⚠️  Integration tests interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        sys.exit(1)
    finally:
        if test_dir:
            cleanup_test_environment(test_dir)

if __name__ == "__main__":
    main()

File: autonomica-frontend/run_integration_tests.js
#!/usr/bin/env node
/**
 * Frontend Integration Test Runner for Autonomica
 * 
 * This script runs integration tests with proper environment setup,
 * Mock Service Worker configuration, and reporting.
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

class IntegrationTestRunner {
  constructor() {
    this.testResults = {
      passed: 0,
      failed: 0,
      total: 0,
      duration: 0
    };
    this.startTime = Date.now();
  }

  log(message, type = 'info') {
    const timestamp = new Date().toISOString();
    const prefix = {
      info: 'ℹ️',
      success: '✅',
      error: '❌',
      warning: '⚠️',
      start: '🚀'
    }[type] || 'ℹ️';
    
    console.log(`${prefix} [${timestamp}] ${message}`);
  }

  checkDependencies() {
    this.log('Checking dependencies...', 'info');
    
    try {
      // Check if Jest is available
      execSync('npx jest --version', { stdio: 'pipe' });
      this.log('Jest is available', 'success');
    } catch (error) {
      this.log('Jest not found. Installing...', 'warning');
      try {
        execSync('npm install --save-dev jest', { stdio: 'pipe' });
        this.log('Jest installed successfully', 'success');
      } catch (installError) {
        this.log('Failed to install Jest', 'error');
        return false;
      }
    }

    try {
      // Check if MSW is available
      execSync('npx msw --version', { stdio: 'pipe' });
      this.log('MSW is available', 'success');
    } catch (error) {
      this.log('MSW not found. Installing...', 'warning');
      try {
        execSync('npm install --save-dev msw', { stdio: 'pipe' });
        this.log('MSW installed successfully', 'success');
      } catch (installError) {
        this.log('Failed to install MSW', 'error');
        return false;
      }
    }

    return true;
  }

  setupTestEnvironment() {
    this.log('Setting up test environment...', 'info');
    
    // Set test environment variables
    process.env.NODE_ENV = 'test';
    process.env.TESTING = 'true';
    process.env.NEXT_PUBLIC_API_URL = 'http://localhost:8000';
    process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY = 'test-key';
    
    this.log('Environment variables set', 'success');
    
    // Create test configuration
    const testConfig = {
      testEnvironment: 'jsdom',
      setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
      testMatch: [
        '<rootDir>/src/**/__tests__/**/*.{js,jsx,ts,tsx}',
        '<rootDir>/src/**/*.{test,spec}.{js,jsx,ts,tsx}'
      ],
      collectCoverageFrom: [
        'src/**/*.{js,jsx,ts,tsx}',
        '!src/**/*.d.ts',
        '!src/**/*.stories.{js,jsx,ts,tsx}',
        '!src/**/*.test.{js,jsx,ts,tsx}',
        '!src/**/index.{js,jsx,ts,tsx}'
      ],
      coverageThreshold: {
        global: {
          branches: 60,
          functions: 60,
          lines: 60,
          statements: 60
        }
      },
      moduleNameMapping: {
        '^@/(.*)$': '<rootDir>/src/$1'
      },
      transform: {
        '^.+\\.(js|jsx|ts|tsx)$': ['babel-jest', { presets: ['next/babel'] }]
      },
      transformIgnorePatterns: [
        '/node_modules/',
        '^.+\\.module\\.(css|sass|scss)$'
      ]
    };
    
    // Write test configuration to file
    const configPath = path.join(__dirname, 'jest.integration.config.js');
    const configContent = `module.exports = ${JSON.stringify(testConfig, null, 2)};`;
    
    try {
      fs.writeFileSync(configPath, configContent);
      this.log('Integration test configuration created', 'success');
    } catch (error) {
      this.log('Failed to create test configuration', 'error');
      return false;
    }
    
    return true;
  }

  runIntegrationTests() {
    this.log('Running integration tests...', 'start');
    
    try {
      const testCommand = 'npx jest --config jest.integration.config.js --testPathPattern=integration --coverage --verbose';
      this.log(`Executing: ${testCommand}`, 'info');
      
      const result = execSync(testCommand, { 
        stdio: 'inherit',
        cwd: __dirname 
      });
      
      this.log('Integration tests completed successfully', 'success');
      return true;
    } catch (error) {
      this.log('Integration tests failed', 'error');
      return false;
    }
  }

  runSpecificIntegrationTests(testPattern = '') {
    this.log(`Running specific integration tests: ${testPattern || 'all'}`, 'info');
    
    try {
      let testCommand = 'npx jest --config jest.integration.config.js --testPathPattern=integration';
      
      if (testPattern) {
        testCommand += ` --testNamePattern="${testPattern}"`;
      }
      
      testCommand += ' --coverage --verbose';
      
      this.log(`Executing: ${testCommand}`, 'info');
      
      const result = execSync(testCommand, { 
        stdio: 'inherit',
        cwd: __dirname 
      });
      
      this.log('Specific integration tests completed successfully', 'success');
      return true;
    } catch (error) {
      this.log('Specific integration tests failed', 'error');
      return false;
    }
  }

  generateTestReport() {
    this.log('Generating test report...', 'info');
    
    const endTime = Date.now();
    const duration = (endTime - this.startTime) / 1000;
    
    console.log('\n📊 Integration Test Report');
    console.log('=' .repeat(50));
    console.log(`⏱️  Total Duration: ${duration.toFixed(2)}s`);
    console.log(`✅ Tests Passed: ${this.testResults.passed}`);
    console.log(`❌ Tests Failed: ${this.testResults.failed}`);
    console.log(`📊 Total Tests: ${this.testResults.total}`);
    
    // Check for coverage reports
    const coverageDir = path.join(__dirname, 'coverage');
    if (fs.existsSync(coverageDir)) {
      console.log('\n📁 Coverage Reports:');
      console.log(`  - HTML: ${path.join(coverageDir, 'lcov-report', 'index.html')}`);
      console.log(`  - LCOV: ${path.join(coverageDir, 'lcov.info')}`);
      console.log(`  - JSON: ${path.join(coverageDir, 'coverage-final.json')}`);
    }
    
    // Check for test results
    const testResultsDir = path.join(__dirname, 'test-results');
    if (fs.existsSync(testResultsDir)) {
      console.log('\n📁 Test Results:');
      const files = fs.readdirSync(testResultsDir);
      files.forEach(file => {
        console.log(`  - ${file}`);
      });
    }
  }

  cleanup() {
    this.log('Cleaning up test environment...', 'info');
    
    try {
      // Remove temporary test configuration
      const configPath = path.join(__dirname, 'jest.integration.config.js');
      if (fs.existsSync(configPath)) {
        fs.unlinkSync(configPath);
        this.log('Test configuration cleaned up', 'success');
      }
      
      // Clean up environment variables
      delete process.env.NODE_ENV;
      delete process.env.TESTING;
      delete process.env.NEXT_PUBLIC_API_URL;
      delete process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY;
      
    } catch (error) {
      this.log('Cleanup failed', 'warning');
    }
  }

  async run() {
    this.log('🚀 Autonomica Frontend Integration Test Runner', 'start');
    console.log('=' .repeat(60));
    
    try {
      // Check dependencies
      if (!this.checkDependencies()) {
        this.log('Dependency check failed', 'error');
        return false;
      }
      
      // Setup test environment
      if (!this.setupTestEnvironment()) {
        this.log('Test environment setup failed', 'error');
        return false;
      }
      
      // Run integration tests
      const success = this.runIntegrationTests();
      
      // Generate report
      this.generateTestReport();
      
      return success;
      
    } catch (error) {
      this.log(`Unexpected error: ${error.message}`, 'error');
      return false;
    } finally {
      this.cleanup();
    }
  }

  async runSpecific(pattern) {
    this.log(`🚀 Running specific integration tests: ${pattern}`, 'start');
    
    try {
      // Setup test environment
      if (!this.setupTestEnvironment()) {
        return false;
      }
      
      // Run specific tests
      const success = this.runSpecificIntegrationTests(pattern);
      
      // Generate report
      this.generateTestReport();
      
      return success;
      
    } catch (error) {
      this.log(`Unexpected error: ${error.message}`, 'error');
      return false;
    } finally {
      this.cleanup();
    }
  }
}

// CLI interface
if (require.main === module) {
  const runner = new IntegrationTestRunner();
  
  const args = process.argv.slice(2);
  const pattern = args[0];
  
  if (pattern && pattern !== '--help') {
    runner.runSpecific(pattern).then(success => {
      process.exit(success ? 0 : 1);
    });
  } else if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Usage: node run_integration_tests.js [pattern] [options]

Options:
  pattern           Run tests matching the specified pattern
  --help, -h       Show this help message

Examples:
  node run_integration_tests.js                    # Run all integration tests
  node run_integration_tests.js "chat"            # Run tests with "chat" in name
  node run_integration_tests.js "authentication"  # Run authentication tests
    `);
    process.exit(0);
  } else {
    runner.run().then(success => {
      process.exit(success ? 0 : 1);
    });
  }
}

module.exports = IntegrationTestRunner;